### ISSUE TITLE
Spack modifies spack.yaml on read-only operations

### ISSUE THREAD

**Issue by jfinney10** [LLNL/spack#23374]
Reference to Spack slack message: https://spackpm.slack.com/archives/C01UZDL35C4/p1619642815329900

When working within an environment where I am adding packages to it, I am getting newlines added to the config file each time I run `spack install`.

### Steps to reproduce the issue

```console
$ spack config edit
...
  definitions:
  - e4s:
    - adios@1.13.1 #+netcdf +szip +fortran +bzip2
    #- adios2@2.6.0 ~bzip2
    - aml@0.1.0
    - arborx@0.9-beta +openmp
    - argobots@1.0
    #- axom@0.3.3
    - bolt@1.0
    #- caliper@2.4.0
...
$ spack install
...
$ spack config edit
...
  definitions:
  - e4s:
    - adios@1.13.1 #+netcdf +szip +fortran +bzip2

    #- adios2@2.6.0 ~bzip2
    - aml@0.1.0
    - arborx@0.9-beta +openmp
    - argobots@1.0

    #- axom@0.3.3
    - bolt@1.0

    #- caliper@2.4.0
```

### System information
OLCF's Ascent

### Spack Version

```
$ spack --version
0.16.1-2391-2473e0c
```

### Additional information

- [ ] I have run `spack debug report` and reported the version of Spack/Python/Platform
Does not apply.
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [ ] I have run the failing commands in debug mode and reported the output
Does not apply

**Comment by jfinney10** [comment]
@scheibelp asked to be pinged on this

**Comment by adamjstewart** [comment]
I've also had this problem in the past, although it's better now than it used to be. Makes it really difficult to keep a `spack.yaml` under version control.

**Comment by eugeneswalker** [comment]
> I've also had this problem in the past, although it's better now than it used to be. Makes it really difficult to keep a `spack.yaml` under version control.

Same boat over here! Would love to see this **not happen** :)

**Comment by alalazo** [comment]
Should be fixed by #37008
